src/asset/asset.cpp

```diff
@@ -0,0 +1,157 @@
+    return NO_ASSET;
+}
+
+AssetData *_raw_fetch(AssetType type, AssetID id) {
+    ASSERT(id < system.num_assets, "Invalid asset id '%lu'", id);
+    ASSERT(system.headers[id].type == type, "Type mismatch, type=%ld, id=%ld", type, id);
+    return &system.data[id];
+}
+
+StringAsset *fetch_string_asset(AssetID id) {
+    return &_raw_fetch(AssetType::STRING, id)->string;
+}
+
+Shader *fetch_shader(AssetID id) {
+    return &_raw_fetch(AssetType::SHADER, id)->shader;
+}
+
+//TODO(gu) re-implement
+template <typename T>
+size_t read(FILE *file, void *ptr, size_t num=1) {
+    if (!num) return 0;
+    size_t read = 0;
+    while (read < num)  {
+        size_t last_read = fread(ptr, sizeof(T), num - read, file);
+        if (!last_read) return 0;
+        read += last_read;
+        ptr = (void *) ((u8 *) ptr + last_read + sizeof(T));
+    }
+    return read;
+}
+
+void load(const char *path) {
+    LOG("Reading asset file %s", path);
+    FILE *file = fopen(path, "rb");
+    if (!file) return;
+
+    LOG("Reading file header");
+    read<FileHeader>(file, &system.file_header, 1);
+    u64 num_assets = system.file_header.num_assets;
+    system.num_assets = num_assets;
+
+    LOG("Reading %d asset headers", num_assets);
+    system.headers = new AssetHeader[num_assets];
+    read<AssetHeader>(file, system.headers, num_assets);
+
+    for (u64 asset = 0; asset < num_assets; asset++) {
+        LOG("Found asset with type %u, name %020lu, data %020lu",
+            system.headers[asset].type,
+            system.headers[asset].name_hash,
+            system.headers[asset].data_hash);
+    }
+
+    system.data = new AssetData[num_assets];
+    for (u64 asset = 0; asset < num_assets; asset++) {
+        AssetHeader header = system.headers[asset];
+        AssetData *data_ptr = &system.data[asset];
+        fseek(file, system.file_header.data_offset + header.data_offset, SEEK_SET);
+        switch (header.type) {
+        case AssetType::TEXTURE: {
+            LOG("Reading Texture at %#lx", ftell(file));
+            read<Image>(file, data_ptr);
+            u64 size = data_ptr->image.size();
+            data_ptr->image.data = new u8[size];
+            LOG("Reading %ld bytes (%dx%dx%d)",
+                size,
+                data_ptr->image.height,
+                data_ptr->image.width,
+                data_ptr->image.channels);
+            read<u8>(file, data_ptr->image.data, size);
+
+            delete[] data_ptr->image.data;
+        } break;
+        case AssetType::STRING: {
+            LOG("Reading String at %#lx", ftell(file));
+            read<StringAsset>(file, data_ptr);
+            u32 size = data_ptr->string.size;
+            data_ptr->string.data = new char[size];
+            LOG("Reading %ld bytes", size);
+            read<char>(file, data_ptr->string.data, size);
+        } break;
+        case AssetType::MODEL: {
+            LOG("Reading Model at %#lx", ftell(file));
```

Comment:
I think you get my point, just remove the logs. I can accept that you want the first one. But the rest are just too verbose.